### DISCUSSION TITLE
[Bug] Theia cloud in https://try.theia-cloud.io/ is not accessible.

### DISCUSSION THREAD

**Discussion by inlann** [eclipse-theia/theia-cloud#318] [Q&A]
When I tried to launch the Theia Blueprint, the page shows an error:

Oh no, something went wrong! ☹
ERROR: Failed to execute 'replace' on 'Location': 'https://' is not a valid URL.

![image](https://github.com/eclipsesource/theia-cloud/assets/33852855/219a37f2-ea96-4270-83d3-b7f0ea3874e2)

**Comment by inlann** [comment]
BTW, I have another question about the deployment of Theia Cloud.

I deployed Theia Cloud in my Kubernetes cluster and noticed that the Operator module creates an Ingress for each user's IDE Pod, resulting in access URLs like 192.168.2.152.nip.io/instance/<uuid> for each IDE Pod. However, when I tried the official demo at https://try.theia-cloud.io/, I noticed that the URL for my IDE Pod did not have the /instance/<uuid> format but rather appeared as xxxx.instance.theia-cloud.io, using a subdomain format.

I observed that my DNS service provider supports wildcard domain configurations like *.test.example.com. I would like to know if the official Theia Cloud setup uses this method, and how I can configure my Theia Cloud deployment to use subdomains in this way.

**Reply by sgraband** [reply]
You can configure the Theia Cloud deployment using subdomains or paths by using these fields: https://github.com/eclipsesource/theia-cloud-helm/blob/ac662ea115ac7341b9e76cda047714a28a353aca/charts/theia-cloud/values.yaml#L80-L93

The comments should give a good overview over what values you need to use.
I can not really give you an answer for the DNS site, as this will be highly custom to your solution, but using wildcards might be one of the solutions.

**Reply by jfaltermeier** [reply]
The values were refactored recently btw. The values of the 0.10.0 release look like this: https://github.com/eclipsesource/theia-cloud-helm/blob/e88973b2253161d8a7a748e1f276bc0ed033a631/charts/theia.cloud/values.yaml#L71

**Reply by inlann** [reply]
> The values were refactored recently btw. The values of the 0.10.0 release look like this: https://github.com/eclipsesource/theia-cloud-helm/blob/e88973b2253161d8a7a748e1f276bc0ed033a631/charts/theia.cloud/values.yaml#L71

Thanks for your reply!  Does the `usePaths` field effect the url of the Service module and Keyclock? Or it only effects the url of users' pod.

**Reply by inlann** [reply]
The cluster of https://try.theia-cloud.io/ seems to be responding again. And the instance url is still using path: `https://ws.theia-cloud.io/06515c94-c310-4f67-925c-3c08345fcc8a/#/home/project`.

I am trying to improve the speed and stability of user access to their IDE Pods, such as by using a CDN to cache static resources like JS and CSS from the IDE containers. This is why I am considering changing the user instance IDs to a subdomain format, which would allow for more precise CDN caching strategies.

**Reply by sgraband** [reply]
Yes `usePaths` effects the service as well, for the keycloak you will need to do the keycloak configuration, which you can do differently if you like.

We will not change the urls of the try.theia-cloud.io deployment, if thats what you are asking.
So to get this straight you want to change your Theia Cloud installation to have these instace urls: `https://<instanceid>.ws.<yourDomain>.com/#/home/project`?

Two considerations here:
- You will not be able to move the path of the current workspace `#/home/project` to the subdomain, due to Theias implementation.
- Moving the instanceId to the subdomain is not a default use case of Theia Cloud at the moment, but is technically feasible. However it would require customizations to the ingresses, service and operator.

**Reply by inlann** [reply]
> * You will not be able to move the path of the current workspace `#/home/project` to the subdomain, due to Theias implementation.
> * Moving the instanceId to the subdomain is not a default use case of Theia Cloud at the moment, but is technically feasible. However it would require customizations to the ingresses, service and operator.

I understand. The workspace information after the # in the URL is not within my scope of concern. I am focused on accelerating the actual network requests for CSS, JS, HTML, and other data seen in the F12 developer tools. These files inherit the IDE pod's URL, such as `https://<theia-cloud-deployment-domain>.com/instances/<uuid>/xx.css` or `https://<theia-cloud-deployment-domain>.com/instances/<uuid>/xx.js` (Currently, I am deploying version 0.8.3.MS1 of theia-cloud).

I know that at the DNS service provider, wildcard domain resolution can be configured to resolve `<instanceid>.<theia-cloud-deployment-domain>.com` to the same IP address. At the CDN service provider, wildcard configurations can also be used to cache static resources like `/*.js` and `/*.css`. These configurations are relatively straightforward.

My main concern is whether it is possible within theia-cloud to set the IDE pod's URL to a format like `<instanceid>.ws.<theia-cloud-deployment-domain>.com`, while also considering the URLs for the landing-page and service modules.

**Reply by sgraband** [reply]
Like i said it will definately be possible, but you will need to do customizations to the code.

**Reply by inlann** [reply]
Thanks, I'll try it.

BTW, can I use my own ssl certificates rather than the cert-manager?

**Reply by jfaltermeier** [reply]
With the 0.11 release that is planed for July (or the current main branches), there will be these helm values:
`ingress.certManagerAnnotations`: `bool`
`ingress.addTLSSecretName`: `bool`
https://github.com/eclipsesource/theia-cloud-helm/tree/main/charts/theia-cloud

You can set `ingress.certManagerAnnotations` to `false` and `ingress.addTLSSecretName` to `true`. Then you may provide the certs via the secret names like this:
https://github.com/eclipsesource/theia-cloud-website/pull/32/files#diff-657dc246a5ef715c9d6c7eaa8bd383c66d2e463855e6b9b951f1c806aa3c0402R128-R132
(this is WIP documentation that will be published with the 0.11 release)

You may also set `ingress.addTLSSecretName` to `false` and patch the ingress controller to use a default certificate that will work for your use case.

**Reply by inlann** [reply]
> Like i said it will definately be possible, but you will need to do customizations to the code.

I tested Gitpod, and it creates a subdomain for each user's pod. However, I found that CDN service providers can configure origin URL rewriting. They can use regex matching to remove the `/instances/<uuid>` string from `/instances/<uuid>/xxx.js`. Our current strategy is to modify the static resource URLs to a fixed format like `cdn.example-domain.com/xxx.js` when packaging the source code and to provide a dedicated static resource server to store these static resource files. Therefore, placing the UUID in the path will not affect our CDN acceleration.

**Reply by inlann** [reply]
> You may also set `ingress.addTLSSecretName` to `false` and patch the ingress controller to use a default certificate that will work for your use case.

That sounds great! Looking forward to the release of version 0.11.

**Comment by sgraband** [comment]
The error on the try-now page should be resolved. Thank you for reporting. We will improve the error messages with our next release.

**Reply by 7ojo** [reply]
Still seeing same error. Does not work for me on MacOS + Firefox

**Reply by sgraband** [reply]
Thanks for reporting. We will check :+1:

**Reply by sgraband** [reply]
We are experiencing unusual load at the moment and the cluster is simply maxed out. We will monitor this and provide a fix asap so that the correct error message is displayed. Thank you for your patience and reporting this.
